shacomp.1
.\" generated with Ronn/v0.7.3
.\" http://github.com/rtomayko/ronn/tree/0.7.3
.
.TH "SHACOMP" "1" "August 2024" "shacomp 1.0" "SHACOMP Manual"
.
.SH "NAME"
\fBshacomp\fR \- Compare SHA (or other) checksums between snapshots
.
.SH "SYNOPSIS"
\fBshacomp\fR [\-N] [\-T] [\-h] [\-H] [\-f] [\-F] [\-u] [\-c ACTIVITY\-CHARS] [\-p PREPROCESSOR] [\-\-help] file file [file \.\.\.]
.
.SH "DESCRIPTION"
\fBshacomp\fR will read multiple files ("snapshots") of SHA (or other) checksums, and display a summary of what files were changed in between each one\.
.
.P
Snapshots are expected to be in a format smilar to that output by \fBsha*sum\fR or \fBmd5sum\fR: A checksum, followed by some number of spaces, followed by the filename, one line per file\.
.
.P
The simplest case is running \fBshacomp\fR on two checksum files, which will show what changes were detected between them\. This will produce a report similar to the following:
.
.IP "" 4
.
.nf

\- /tmp/sortELfgNZ
\- /tmp/sortL93Y32
+ /tmp/sortdxUTk2
+ /tmp/sorte9MJK5
# /var/lib/mlocate/mlocate\.db
# /var/lib/mysql/ib_logfile0
.
.fi
.
.IP "" 0
.
.P
Each file is preceeded by a symbol which shows the nature of the change:
.
.IP "\(bu" 4
A \fB+\fR indicates an "addition": It was present in the older digest, but not the newer\.
.
.IP "\(bu" 4
A \fB\-\fR indicates a "deletion": It was present in the newer digest, but not the older\.
.
.IP "\(bu" 4
A \fB#\fR indicates a "modification": It was present in both digests, but with a differing checksum\.
.
.IP "" 0
.
.P
By default, files which are "unchanged" \-\- they exist in both digests, with the same checksum \-\- are not listed\.
.
.P
\fBshacomp\fR can also compare multiple digest files\. If more than two files are provided, it will track changes across all of them, giving a more detailed report:
.
.IP "" 4
.
.nf

+\-\-telos\-digest\-1722816420\-Aug04\-05\.07pm\-0700\.sha
|+\-\-telos\-digest\-1722830639\-Aug04\-09\.03pm\-0700\.sha
||+\-\-telos\-digest\-1722926516\-Aug05\-11\.41pm\-0700\.sha
|||+\-\-telos\-digest\-1723064401\-Aug07\-02\.00pm\-0700\.sha
||||+\-\-telos\-digest\-1723150801\-Aug08\-02\.00pm\-0700\.sha
|||||+\-\-telos\-digest\-1723237201\-Aug09\-02\.00pm\-0700\.sha
||||||
+|#||| /etc/passwd
+##||| /etc/shadow
|+\-||| /tmp/sort4hPmeO
||||+| /var/spool/postfix/maildrop/0418E2C0C1A
.
.fi
.
.IP "" 0
.
.P
The first few lines lists the checksum files provided\. Following that is a list of all files that were changed at any point, preceeded by a set of symbols showing where changes were detected\.
.
.P
Each symbol represents one of the checksum files; it shows what changed in \fIthat\fR snapshot\. I\.E\. what changed between that snapshot and the \fIprevious\fR snapshot\.
.
.P
The set of symbols are the same as in the first example, with one addition: A \fB|\fR indicates that a file was unchange between snapshots\. I\.E\. Either it appeared in both files, with the same checksum, or did not appear in either file\.
.
.P
Note that this report includes a column for the \fIfirst\fR snapshot\. Since there is no "previous" snapshot to compare to (and thus, no meaningful possibility of "removal" or "modification"), this column will always contain either a \fB+\fR (if the file was present in the first snapshot), or a \fB|\fR (if it was not)\. This makes it slighly easier to see which snapshot a file first appeared in, and makes the display overall more consistent\.
.
.P
Similar to the first example, a file will not be listed if it was unchanged across \fIall\fR snapshots\.
.
.SH "OPTIONS"
.
.TP
\fB\-N\fR, \fB\-\-newest\-first\fR
In order to generate a meaningful report, \fBshacomp\fR needs to know the chronological order of the checksum snapshots; otherwise, the "when" of file changes would get mixed up\. By default, \fBshacomp\fR assumes that the files given on the command line are in chronological order, from oldest to newest\. This option causes \fBshacomp\fR to assume they are provided in the reverse order\.
.
.TP
\fB\-T\fR, \fB\-\-check\-timestamps\fR
This option causes \fBshacomp\fR to check the timestamps of the snapshot files to see what order they should be processed in\. In this mode, it does not matter what order they are specified on the command line\.
.
.TP
\fB\-h\fR, \fB\-\-header\fR
Display a header showing the given snapshot files\. The default if more than 2 files are given\.
.
.TP
\fB\-H\fR, \fB\-\-no\-header\fR
Do not show the header
.
.TP
\fB\-f\fR, \fB\-\-first\-file\-changes\fR
Displays a column for the first snapshot\. The default if more than 2 files are given\.
.
.TP
\fB\-F\fR, \fB\-\-no\-first\-file\-changes\fR
Do not dispay a column for the first snapsnot\.
.
.TP
\fB\-u\fR, \fB\-\-show\-unchanged\fR
Include files that were unchanged across all snapshots
.
.TP
\fB\-c\fR, \fB\-\-activity\-chars\fR
Specify the symbols used to indicate each type of file change\. This should be a 4 character string where the symbol for unchanged, added, deleted, and modified files should be specified, in that order\. Default is "|+\-#"\.
.
.TP
\fB\-p\fR, \fB\-\-preprocessor\fR
When reading the snapshot files, first pass their contents through the given program\. The content of the file will be sent to the program on stdin, and the result will be read in from stdout\. This is useful for, say, decompressing compressed snapshots on\-the\-fly by passing them through \fBgunzip\fR or \fBunxz\fR or similar\. The argument is passed through the system shell, so a pipeline or other shell construct may be given\.
.
.SH "NOTES"
\fBshacomp\fR is written in Python\. It targets Python 3\.9, and thus will run using pypy\.
.
.P
This manual page was created using ronn(1)\.
.
.SH "HISTORY"
.
.TP
1\.0
Initial version
.
.TP
1\.1
Added \fB\-\-preprocessor\fR option
.
.SH "COPYRIGHT"
\fBshacomp\fR is Copyright 2024 Nathan Roberts \fInroberts@tardislabs\.com\fR\.
.
.P
This project is licensed under the terms of the MIT license\.


shacomp.py
#!/usr/bin/env python3

import argparse
from argparse import ArgumentParser
from pathlib import Path
from pprint import pp
from sys import stderr
from subprocess import Popen, PIPE

def main():
    parser = ArgumentParser(
            add_help=False,
            description='Compare SHA digests'
            )

    parser.add_argument( '-N', '--newest-first', dest='forder',
                        action='store_const', const='newest',
                        default='oldest',
                        help='Assume given digests provided in order from '
                        'newest to oldest. (Default is oldest-to-newest)' )
    parser.add_argument( '-T', '--check-timestamps', dest='forder',
                        action='store_const', const='timestamp',
                        help='Check digest file timestamps to determine order' )
    parser.add_argument( '-h', '--header', action='store_true', default=None,
                        help='Display header (default if more than 2 files)' )
    parser.add_argument( '-H', '--no-header', action='store_false',
                        dest='header',
                        help='Do not display header' )
    parser.add_argument( '-f', '--first-file-changes', action='store_true',
                        dest='first', default=None,
                        help='Display change column for first digest file ' +
                        '(default if more than 2 files)' )
    parser.add_argument( '-F', '--no-first-file-changes', action='store_false',
                        dest='first',
                        help="Don't display change column for first "
                        "digest file" )
    parser.add_argument( '-u', '--show-unchanged', action='store_true',
                        help="Include files that were unchanged across "
                        "all digests" )
    parser.add_argument( '-c', '--activity-chars', default='|+-#',
                        help='Characters to display for files that are '
                        'unchanged, added, removed, or modified. '
                        '(Default "|+-#")' )
    parser.add_argument( '-p', '--preprocessor', 
                        help='Send the content of each snapshot file through '
                        'PREPROCESSOR. (Eg. gunzip, unxz, etc.)' )
    parser.add_argument( '--help', action='help',
                        help='Show this help message' )

    parser.add_argument( 'ffile', type=Path, metavar='file',
                        help=argparse.SUPPRESS )
    parser.add_argument( 'files', nargs='+', type=Path, metavar='file',
                        help='Digest file to compare' )

    args = parser.parse_args()

    if len( args.files ) > 1:
        if args.first == None: args.first = True
        if args.header == None: args.header = True

    if len( args.activity_chars ) != 4:
        print( "ACTIVITY_CHARS must be exactly 4 characteres.", file=stderr )
        exit( 1 )

    if 0:
        print( args )
        exit()

    # List of digest files, in chronological order
    digestFiles = []
    digestFiles.append( args.ffile )
    digestFiles.extend( args.files )

    # Reverse files if they were given in reverse order
    if args.forder == 'newest':
        digestFiles.reverse()
    elif args.forder == 'timestamp':
        for f in digestFiles:
            digestFiles.sort( key=lambda f: f.stat().st_mtime )

    if 0:
        for f in digestFiles:
            print( str( f ))
        exit()

    # Set of files in all digests
    allFiles = set()
    # Data in each digest file (in same order as digestFiles)
    digests = []

    # Read all the digest files
    for digestFile in digestFiles:
        with open( digestFile ) as f:
            if args.preprocessor:
                with Popen( args.preprocessor, shell=True,
                stdin=f, stdout=PIPE ) as proc:
                    data = readDigests( proc.stdout )
            else:
                data = readDigests( f )
        digests.append( data )
        allFiles |= set( data.keys() )

    # Sort file list
    files = list( allFiles )
    files.sort()

    # Print header
    if args.header:
        # Extra number of pipes to print per line
        offs = 0
        # What to print before first file
        ff = ". "
        if args.first:
            offs = 1
            ff = "+--"

        print( f"{ff}{str( digestFiles[0] )}" )
        for i in range( 1, len( digestFiles )):
            print( f"{'|' * (i + offs - 1)}+--{str( digestFiles[i] )}" )
        print( "|" * (len( digestFiles ) + offs - 1) )

    # Go through each file and find changes between digests
    for file in files:
        # Were changes found in any revision?
        anyChange = False
        # Status of each file for each rev
        stat = ""

        if args.first:
            if file in digests[0]:
                stat = args.activity_chars[1]
            else:
                stat = args.activity_chars[0]

        # Go through each digest and compare to the next one
        for i in range( len( digests ) - 1 ):
            old = digests[i]
            new = digests[i + 1]
            
            if file in old and file not in new:
                # File removed from later dataset
                stat += args.activity_chars[2]
                anyChange = True
            elif file in new and file not in old:
                # File added to later dataset
                stat += args.activity_chars[1]
                anyChange = True
            elif file in old and file in new and old[file] != new[file]:
                # File changed between digests
                stat += args.activity_chars[3]
                anyChange = True
            else:
                # No change between digests
                stat += args.activity_chars[0]

        if anyChange or args.show_unchanged:
            print( f"{stat} {file}" )

def readDigests( fd ):
    """Read a snapshot file from the given file object.

    Returns a dict with filenames as keys and checksums as values.
    """
    data = {}
    for line in fd.readlines():
        if isinstance( line, bytes ):
            line = line.decode()
        (sha, filename) = line.split( ' ', maxsplit=1 )
        filename = filename.lstrip( ' ' ).rstrip( "\n" )
        data[filename] = sha
    return data

if __name__ == "__main__":
    main()



